Fixes #19155 Editable anchors for glissandi and other note-anchored spanners

mgavioli opened the pull request:
Fixes #19155, #22861 (duplicate of the former) and #23100.

__References__:
Issues:	https://musescore.org/en/node/19155 https://musescore.org/en/node/22861 https://musescore.org/en/node/23100

__Description__:
Allows to change the start and end note to which a glissando is anchored after it has been entered. Either anchor can be changed independently.

The user interface follows the current working of other 'snappable' lines. Once either the start or end grip is selected:
- `[Shift]+[Left]` snaps the anchor to the previous chord, defaulting to its top note.
- `[Shift]+[Right]` snaps to the next chord, defaulting to its top note.
- `[Shift]+[Up]` snaps to the note above (possibly in a chord, voice or staff above the current one).
- `[Shift]+[Down]` snaps to the note below (possibly in a chord, voice or staff below the current one).

This permits to set the anchor points of a glissando to any note in the score, allowing several glissandi between the notes of the same two chords and other complex configurations (glissandi skipping intermediate chords, start and end notes in different voices or staves, and so on).

It is possible to move the anchor to a different staff of the same instrument, but not to a different instrument; also, it is not possible to 'cross' a change of instrument in the same staff.

__Known limitations__:
- The `[Shift]+[Up]` and `[Shift]+[Down]` use the same note-finding functions as the `[Alt]+[Up]` and `[Alt]+[Down]`actions which move the selection cursor to the above and below note, even across voices or staves (namely `Score::upAlt()` and `Score::downAlt()`). Occasionally, in particular if the note immediately above or below is not time-aligned, the algorithm has little expected results; however, the behaviour is already known to the user. Improving the algorithm would benefit both uses.

__Notes__:
- Most of the added infrastructure is not specific to glissando but to any spanner anchored to notes, then it should also add after-the-fact "snap to" note support to note-anchored text line.
- When moving an anchor, the algorithm usually prefers a note in the same voice/staff of the old note if it exists; if there is none, it tries other voices of the same staff.
- The change of anchor is undoable.
- The fix corrects the management of the `Chord::_endsGlissando` flag, taking into account that a chord can be the ending point of several glissandi and removing one of them not necessarily means the chord no longer ends a glissando (another glissando may still exists).
- The fix also improved the rendering of the glissando wavy line, with better alignment with anchor notes and, with glissando text, better text-line spacing.

lasconic commented:
I'm testing this PR, on top of the current master fdb68b6cbdd982fed77d9316d3a4f5d686fa5bde

I create score with the SATB + piano and add a couple of notes in the first two staves. Select a note in the second staff and double click a glissando in the palette. Double click the glissando in the score. The anchor looks wrong. If I use Shift + Down several times, it crashes.

<img width="317" src="https://cloud.githubusercontent.com/assets/234271/9521402/070c1f2a-4ccf-11e5-9b1c-eb66e3746c1d.png">

mgavioli commented:
@lasconic : Fixed (hopefully) both the strange anchor display and the crash when [Shift]-moving to a non-note.

I had to change the return value of `SLine::linePos()` for note-anchored lines, generalizing the values previously returned for glissandi; but, as far as I know, glissandi are currently the only note-anchored spanners (the menu command to add a note-anchored text line seems to do nothing). So, it should not affect anything else

MarcSabatella commented:
Actually, the note anchored textline *is* a real thing.  You have to select two notes first, then use the menu.  A line will be drawn between them, at an offset of 5sp above.  I don't think it is particularly well-known, used, or tested.

mgavioli commented:
@MarcSabatella : note-anchored text lines do exist indeed! Thanks for the explanation. However, in 2.0.2 (and, I think, earlier) they crash MS as soon as one attempts to move the anchors.

My PR fixes also this crash, but its current version displaces the note-anchored text lines vertically (staff-relative vs. system-relative coordinates disagreement). I'm fixing this and I'll add a new commit.

mgavioli commented:
@lasconic @MarcSabatella : note-anchored text lines should work now: everything working before should work the same and, in addition, anchors can be changed to other notes of the same staff or of a different staff.

lasconic commented on libmscore/utils.cpp:
```diff
@@ -290,6 +290,62 @@ Segment* prevSeg1(Segment* seg, int& track)
+      // TODO : limit to same instrument, not simply to same staff!
+      Segment*    seg   = note->chord()->segment()->prev1();
+      while (seg) {
+            if (seg->segmentType() != Segment::Type::ChordRest)
```
If seg is not a Segment::Type::ChordRest when we enter the loop? Infinite loop :boom:

mgavioli commented:
@lasconic : and the last line of that function should be `seg = seg->prev1();` not `seg = seg->nextCR(track, true);` damn cut-n-paste! Fixed

lasconic commented on libmscore/line.cpp:
```diff
@@ -205,7 +209,9 @@ bool LineSegment::edit(MuseScoreView* sv, Grip curGrip, int key, Qt::KeyboardMod
       int track2  = l->track2();    // assumed to be same as track
 
-      if (l->anchor() == Spanner::Anchor::SEGMENT) {
+      switch(l->anchor()) {
```
Can you check the indenting here?

mgavioli commented:
Fixed the points discussed on IRC:
- The `for` loop at: https://github.com/musescore/MuseScore/pull/2155/files#diff-428e0bb12d899d2d494b3cc99e75e22fR2995
- The indenting of `switch`es at https://github.com/musescore/MuseScore/pull/2155/files#diff-97d139714d630a8f95d269d78b4d8fb1R213 and following
- Blocked the move of anchors to different instruments: https://github.com/musescore/MuseScore/pull/2155/files#diff-97d139714d630a8f95d269d78b4d8fb1R213 This also blocks a glissando 'across' an instrument change in the same staff.

Squashed all the commits into one.

CombatCube commented:
I can no longer edit glissandi with this PR. Upon placing one into my score, I can select it but double-click has no effect. The "Edit Element" option in the context menu is greyed out.

Running on Windows 8.1 on commit d45cc192e5fe7b1c33f71ed2cd43f9140aa72312

MarcSabatella commented:
Works fine for me on Ubuntu using a current development build (with PR already merged).  Might be something specific to your score or the exact sequence of operations?
